### PR TITLE
Remove uses of DeprecatedString in file utility.

### PR DESCRIPTION
**Pull request by vkoskiv**
Found out that most of it can just be `StringView` anyway, because the source strings are static byte arrays.

### CODE REVIEW

File: Userland/Utilities/file.cpp [hunk 1]
```diff
@@ -19,59 +20,59 @@
...
     if (!image_decoder)
-        return {};
+        return OptionalNone {};
 
-    return DeprecatedString::formatted("{}, {} x {}", description, image_decoder->width(), image_decoder->height());
+    return TRY(String::formatted("{}, {} x {}", description, image_decoder->width(), image_decoder->height()));
 }
 
-static Optional<DeprecatedString> gzip_details(DeprecatedString description, DeprecatedString const& path)
+static ErrorOr<Optional<String>> gzip_details(StringView const& description, StringView const& path)
```
Comment (MacDue) [inline Userland/Utilities/file.cpp:44]: And here:
```suggestion
static ErrorOr<Optional<String>> gzip_details(StringView description, StringView path)
```

File: Userland/Libraries/LibCore/MimeData.cpp [hunk 2]
```diff
@@ -112,70 +112,70 @@ StringView guess_mime_type_based_on_filename(StringView path)
...
-Optional<DeprecatedString> guess_mime_type_based_on_sniffed_bytes(ReadonlyBytes bytes)
+Optional<StringView> guess_mime_type_based_on_sniffed_bytes(ReadonlyBytes bytes)
 {
 #define __ENUMERATE_MIME_TYPE_HEADER(var_name, mime_type, pattern_offset, pattern_size, ...)                       \
     if (static_cast<ssize_t>(bytes.size()) >= pattern_offset && bytes.slice(pattern_offset).starts_with(var_name)) \
         return mime_type;
     ENUMERATE_HEADER_CONTENTS;
 #undef __ENUMERATE_MIME_TYPE_HEADER
-    return {};
+    return OptionalNone {};
```
Comment (linusg) [inline Userland/Libraries/LibCore/MimeData.cpp:179]: This shouldn't be needed (you didn't change `mime_type_for_file()` either)

**Review by MacDue — COMMENTED** [review]
Some nit picks:

File: Userland/Utilities/file.cpp [hunk 3]
```diff
@@ -19,59 +20,59 @@
 #include <sys/stat.h>
 #include <unistd.h>
 
-static Optional<DeprecatedString> description_only(DeprecatedString description, [[maybe_unused]] DeprecatedString const& path)
+static ErrorOr<Optional<String>> description_only(StringView const& description, [[maybe_unused]] StringView const& path)
```
Comment (linusg) [inline Userland/Utilities/file.cpp:23]: Never pass string views by const reference, always by value

File: Userland/Utilities/file.cpp [hunk 4]
```diff
@@ -19,59 +20,59 @@
...
 {
-    return description;
+    return String::from_utf8(description);
 }
 
 // FIXME: Ideally Gfx::ImageDecoder could tell us the image type directly.
-static Optional<DeprecatedString> image_details(DeprecatedString const& description, DeprecatedString const& path)
+static ErrorOr<Optional<String>> image_details(StringView const& description, StringView const& path)
```
Comment (linusg) [inline Userland/Utilities/file.cpp:29]: Same here, and below

File: Userland/Utilities/file.cpp [hunk 5]
```diff
@@ -80,71 +81,71 @@ static Optional<DeprecatedString> elf_details(DeprecatedString description, Depr
...
+    return TRY(String::formatted("{} {}-bit {} {}, {}, version {} ({}){}",
         description,
         bitness,
         byteorder,
         ELF::Image::object_file_type_to_string(header.e_type).value_or("(?)"sv),
         ELF::Image::object_machine_type_to_string(header.e_machine).value_or("(?)"sv),
         header.e_ident[EI_ABIVERSION],
         ELF::Image::object_abi_type_to_string(header.e_ident[EI_OSABI]).value_or("(?)"sv),
-        is_dynamically_linked ? dynamic_section : "");
+        is_dynamically_linked ? dynamic_section : String()));
```
Comment (linusg) [inline Userland/Utilities/file.cpp:94]: `String {}` for consistency with other code using the new empty string

**Review by linusg — CHANGES_REQUESTED** [review]
Missing punctuation in first commit message:

https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md#human-language-policy
> The following applies to all user-facing strings, code, comments, and commit messages:
> - Use proper spelling, grammar, and punctuation.

File: Userland/Utilities/file.cpp (hunk 3 above)
Comment (linusg) [reply Userland/Utilities/file.cpp:23]: For reference: https://github.com/SerenityOS/serenity/commit/8b1108e4858f797c9216dc8ae4a3918ad50c73b4

**Comment by vkoskiv** [comment]
Thanks for the good feedback @MacDue and @linusg !

Fixups amended.

File: Userland/Utilities/file.cpp [hunk 6]
```diff
@@ -19,59 +20,59 @@
...
 // FIXME: Ideally Gfx::ImageDecoder could tell us the image type directly.
-static Optional<DeprecatedString> image_details(DeprecatedString const& description, DeprecatedString const& path)
+static ErrorOr<Optional<String>> image_details(StringView description, StringView path)
 {
     auto file_or_error = Core::MappedFile::map(path);
     if (file_or_error.is_error())
-        return {};
+        return OptionalNone {};
 
     auto& mapped_file = *file_or_error.value();
```
Comment (trflynn89) [inline Userland/Utilities/file.cpp:35]: Now that we're returning ErrorOr, this feels like an error we should propagate.
```c++
auto mapped_file = TRY(Core::MappedFile::map(path));
```

File: Userland/Utilities/file.cpp [hunk 7]
```diff
@@ -19,59 +20,59 @@
...
+        return OptionalNone {};
 
-    return DeprecatedString::formatted("{}, {} x {}", description, image_decoder->width(), image_decoder->height());
+    return TRY(String::formatted("{}, {} x {}", description, image_decoder->width(), image_decoder->height()));
 }
 
-static Optional<DeprecatedString> gzip_details(DeprecatedString description, DeprecatedString const& path)
+static ErrorOr<Optional<String>> gzip_details(StringView description, StringView path)
 {
     auto file_or_error = Core::MappedFile::map(path);
```
Comment (trflynn89) [inline Userland/Utilities/file.cpp:46]: Same here

File: Userland/Utilities/file.cpp [hunk 8]
```diff
@@ -19,59 +20,59 @@
...
+        return OptionalNone {};
 
-    return DeprecatedString::formatted("{}, {}", description, gzip_details.value());
+    return TRY(String::formatted("{}, {}", description, gzip_details.value()));
 }
 
-static Optional<DeprecatedString> elf_details(DeprecatedString description, DeprecatedString const& path)
+static ErrorOr<Optional<String>> elf_details(StringView description, StringView path)
 {
     auto file_or_error = Core::MappedFile::map(path);
```
Comment (trflynn89) [inline Userland/Utilities/file.cpp:63]: Same here

File: Userland/Utilities/file.cpp [hunk 9]
```diff
@@ -19,59 +20,59 @@
...
+        return OptionalNone {};
     auto& mapped_file = *file_or_error.value();
     auto elf_data = mapped_file.bytes();
     ELF::Image elf_image(elf_data);
     if (!elf_image.is_valid())
-        return {};
+        return OptionalNone {};
 
     StringBuilder interpreter_path_builder;
     auto result_or_error = ELF::validate_program_headers(*(const ElfW(Ehdr)*)elf_data.data(), elf_data.size(), elf_data, &interpreter_path_builder);
```
Comment (trflynn89) [inline Userland/Utilities/file.cpp:73]: Same here

**Review by trflynn89 — COMMENTED** [review]
Seems like there's several errors that should be propagated now that some of the handlers return ErrorOr (feel free to do that as a separate commit or PR even).

**Comment by vkoskiv** [comment]
Done, errors now propagate from file detail provider functions. Reduced lines of code!

**Comment by trflynn89** [comment]
So I think the first commit isn't needed (or rather isn't the right fix). We already define

**Comment by trflynn89** [comment]
So I think the first commit isn't needed (or rather isn't the right fix). We already define this HashCompatible override in StringUtils.h:
https://github.com/SerenityOS/serenity/blob/9a2ee5a9dd792359440429832ac564f5dde2b315/AK/StringUtils.h#L17-L18

That *should* work for `String` as well. The problem is it doesn't work at all and will need to be fixed separately:
https://discord.com/channels/830522505605283862/860165510263996427/1070727121225388152

**Comment by vkoskiv** [comment]
> So I think the first commit isn't needed (or rather isn't the right fix). We already define this HashCompatible override in StringUtils.h:
> 
> https://github.com/SerenityOS/serenity/blob/9a2ee5a9dd792359440429832ac564f5dde2b315/AK/StringUtils.h#L17-L18
> 
> That _should_ work for `String` as well. The problem is it doesn't work at all and will need to be fixed separately: https://discord.com/channels/830522505605283862/860165510263996427/1070727121225388152

This PR won't compile without it, so it cannot be merged without it.

**Comment by trflynn89** [comment]
> This PR won't compile without it, so it cannot be merged without it.

I know - I'm working on a fix. But we also shouldn't merge this PR *with* it because it isn't right (which I understand is unfortunate and annoying).

**Comment by vkoskiv** [comment]
OK

**Comment by trflynn89** [comment]
You can leave the PR open if you want :sweat_smile: I would just mark it a draft or add a blocked label

**Comment by vkoskiv** [comment]
@trflynn89 Dropped the AK commit, this is now blocked on the appropriate fix.

**Comment by vkoskiv** [comment]
I can't seem to add a "Blocked" label, though!

**Comment by vkoskiv** [comment]
Thanks :D

**Pull request closed without merging**